### PR TITLE
Use placeholder tensor in scan

### LINKED ISSUE #8612: Placeholder tensors
## 🚀 Feature

**This proposal is to create a version of `torch.empty(...)` tensor that does not allocate memory on the accelerator, here called a placeholder tensor. If you read the data of this tensor, it will crash. But it should be possible to abstractly stage out an HLO from an IR graph by feeding this tensor as input.**

## Motivation

When comparing the metrics of a model run using for loop vs run using `scan`, it appears that the scan version transfers a lot more data to the device. From the metrics, it transfers 23 GiB compared to 182 MiB when using for loops. The metric `IrValueTensorToXlaData` has a large count (>200) under scan compared to 0 under for loop. The profile timeline also shows additional tensors that are sent to the TPU during LazyTensor tracing.

When I list out the operations that cause a tensor to be sent to the device, specifically via the IrValueTensorToXlaData path, it appears that any graph operation that touches an input tensor (e.g. primals, tangents) cause a tensor to be sent to the device: see [3]. For example, the `t = torch.ops.aten.t.default(primals_2)` operation transposes `primal_2`, and that likely causes the data backing primal_2 to be sent to the device.

I have narrowed this down to the `make_fake_tensor` [4] function in scan, which creates device data inputs that are used to trace the combine function:

```
  def make_fake_tensor(v: torch.Tensor) -> torch.Tensor:
    return torch.empty(
        v.size(), dtype=v.dtype).to(device).requires_grad_(v.requires_grad)
```

I believe this function associates a CPU tensor with the result as its backing data somehow, but the resulting tensor does not have an IR value. Later, when some operation (such as torch.ops.aten.t) uses the tensor, its IR value is required. PyTorch/XLA then creates the IR value by transferring the data tensor to the device and then minting a DeviceData IR node referencing the data tensor. This behavior is documented in Virtual Devices for GSPMD on PyTorch/XLA [1].

The end result is that even if we later throw away all the outputs of the combine function after extracting its HLO, simply creating the inputs to run the combine function still has the side-effect of sending lots of empty tensors to the device. The size of these tensors do not appear to be zero despite what its name suggests. This slows down training and should be avoided. For now I'll attempt to workaround this by caching the HLO computation corresponding to the combine function. At least this way we only trace and send unnecessary data once, after which the HLO is reused in subsequent scan invocations involving the same combine function.

This is not a problem in Jax since scan in Jax traces the combine function abstractly i.e. abstract tensors are passed to the function instead of concrete tensors holding device data. PyTorch/XLA currently doesn't have the concept of an abstract tensor, so all lazy-tensor based tracing operations are prone to generating side-effects from accidentally sending data to the device to accidentally materializing the tensor.

In Llama 405B, the extra transfers become significant enough to OOM the TPU just from staging out the HLO of the combine function in scan. We'll need to find a representation of these input tensors used during tracing such that they are lowered into HLO parameters but don't transfer data to the device.

## Pitch

On Dec 20, 2024, I got some suggestions from jackcao@ where we could create a tensor without a data handle, also called a "placeholder". That tensor can still be lowered and it won't eagerly allocate buffers on the device. When I implemented this suggestion, it looks like LoweringContext is using the data handle to deduplicate parameters for some reason, so it appears that we still need the data handle or at least some kind of unique identifier. For now I worked around this by [falling back](https://github.com/pytorch/xla/commit/8158d947f80b1c787e290f42b9666fe477f2156d) to the `torch::lazy::BackendData` pointer when the BackendData object does not hold a handle. Unknown if this approach has holes. It probably does and needs to be properly fixed.

[1]: https://docs.google.com/document/d/1GmT-os8xyb30kuwYm53WvxV3LtI5Ag8uaJCMWAlVdWE/edit?tab=t.0
[3]: https://docs.google.com/document/d/1ImNqqAJV_XVFlrcneuRrtKELw6NOsJQXCfIqmXXCyB8/edit?usp=sharing
[4]: https://github.com/pytorch/xla/blob/v2.6.0-rc1/torch_xla/experimental/scan.py#L468-L470

### PR DESCRIPTION
**Pull request by tengyifei**
Fixes #8612. See that issue for the motivation.

This PR introduces a `_xla_create_placeholder_tensor` function that creates a `DeviceData` lazy tensor that does not contain a buffer. It then uses that in `scan` when staging out the HLO of a function so that we don't send buffers to the TPU while tracing.

This PR replaces a `XLA_CHECK(HasValue())` fail with `return reinterpret_cast<uintptr_t>(this);`. Because the previous behavior is to crash, this PR should not break any existing code. Still, I've tried to understand what is the meaning of `torch::lazy::BackendData::Handle`, with the result of the investigation documented below.

----

### What is a `torch::lazy::BackendData::Handle`?

The best documentation I can find is [1] which says "represents tensor data stored on a backend device". That doesn't say much.

Note that even PyTorch is confused what is a "Handle". You would think the `torch::lazy::LazyTensor::CurrentDataHandle()` function returns a `torch::lazy::BackendData::Handle`, but it instead returns a `torch::lazy::BackendDataPtr`, which is a shared pointer around a `torch::lazy::BackendData`. So you can end up with code like `auto handle = tensor->CurrentDataHandle(); auto handlehandle = handle->GetHandle();`. If we see some variable named "handle", we need to check whether it's a `std::shared_ptr<torch::lazy::BackendData>` or `torch::lazy::BackendData::Handle`.

In summary, there's no rationalized explanation of what is a handle I could find. Therefore I'll infer its semantics from the usages.

### What are the use sites of `BackendData::GetHandle`?

- `torch_xla._XLAC._get_tensors_handle`: only used in one test
- `torch_xla._XLAC._get_tensors_xla_device_data_node`: uses it to deduplicate DeviceData nodes.
- `PyLoweringContext::GetTensorParameterId` and `LoweringContext::GetParameter`: checks if a tensor is an existing parameter in the lowering context. Also used to deduplicate HLO parameters based on the `Handle` value of tensors referenced in a graph. That change is introduced in [2], in a PR merged without any comments.

As far as I can tell, it's used to deduplicate device data tensors, so we need to ensure that when we create two placeholder tensors with `_xla_create_placeholder_tensor`, they'll still have different handles. The easiest way is to use the address of the `BackendData` object.

[1]: https://github.com/pytorch/pytorch/pull/67927/files
[2]: https://github.com/pytorch/xla/pull/1080/files

### CODE REVIEW

File: torch_xla/csrc/runtime/pjrt_computation_client.h [hunk 1]
```diff
@@ -191,6 +191,12 @@ class PjRtComputationClient : public ComputationClient {
...
     Handle GetHandle() override {
+      // If the data is a placeholder, use the address of this object as the
+      // handle.
+      if (buffer == nullptr) {
+        return reinterpret_cast<std::uintptr_t>(this);
+      }
+
       XLA_CHECK(HasValue())
           << "buffer with shape " << shape().ToString() << " on device "
           << device() << (buffer == nullptr ? " is null" : " is deleted");
```
Comment (rpsilva-aws) [inline torch_xla/csrc/runtime/pjrt_computation_client.h:202]: Maybe this should this be buffer-IsDeleted(), if that's the accurate deletion state for this PjRtBuffer.

**Review by rpsilva-aws — COMMENTED** [review]
Nice! Thanks for adding some context behind the different handles.

File: torch_xla/csrc/runtime/ifrt_computation_client.h [hunk 2]
```diff
@@ -203,6 +203,12 @@ class IfrtComputationClient : public ComputationClient {
           sharding_(sharding) {}
 
     Handle GetHandle() override {
+      // If the data is a placeholder, use the address of this object as the
```
[the remark this answers is not in the document]
Comment (tengyifei) [reply torch_xla/csrc/runtime/ifrt_computation_client.h:206]: I removed the redundant XLA_CHECK. FWIW, AFAIK the IFRT client is completely unused. So whatever I do here is pretty moot. But I think your point applies to PJRT as well.

File: torch_xla/csrc/runtime/ifrt_computation_client.h [hunk 3]
```diff
@@ -203,6 +203,12 @@ class IfrtComputationClient : public ComputationClient {
...
+      // handle.
+      if (buffer == nullptr) {
```
[the remark this answers is not in the document]
Comment (tengyifei) [reply torch_xla/csrc/runtime/ifrt_computation_client.h:208]: No longer relevant

File: torch_xla/csrc/runtime/ifrt_computation_client.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (tengyifei) [reply torch_xla/csrc/runtime/ifrt_computation_client.h:206]: I'm going to test that deleting a PJRT data actually leads to the Deleted state

File: torch_xla/csrc/init_python_bindings.cpp [hunk 4]
```diff
@@ -1876,6 +1877,18 @@ void InitXlaModuleBindings(py::module m) {
           XLATensorPtr output = tensor_methods::einsum(equation, xla_tensors);
           return bridge::AtenFromXlaTensor(output);
         });
+
+  // Creates a placeholder tensor that does not hold any device buffer.
+  // This is primarily useful for staging out the HLO of a user computation.
+  // Accessing the value of the tensor will panic.
```
[the remark this answers is not in the document]
Comment (tengyifei) [reply torch_xla/csrc/init_python_bindings.cpp:1883]: Done

File: torch_xla/experimental/scan.py [hunk 5]
```diff
@@ -558,8 +558,10 @@ def fn(carry, x):
   # Abstractly trace and lower `fn`.
   # Later we will include `fn_computation` within the while loop body.
   def make_fake_tensor(v: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (tengyifei) [reply torch_xla/experimental/scan.py:560]: github doesn't like stacking prs and i'm in a hurry :)

File: test/scan/test_scan.py [hunk 6]
```diff
@@ -613,6 +613,32 @@ def compute_outputs_and_gradients(carry, x):
...
+                      requires_grad=True,
+                      device=self.device)
+    torch_xla.sync(wait=True)
+    met.clear_all()
+    self.assertFalse(met.metric_data("TransferToDeviceTime"))
+    # Use `scan` to lower `fn` into HLO and run it. Doing so should not
+    # transfer anything from host to device since `init` and `xs` are
+    # already on the device.
+    # In practice, `carry` and `x` will be placeholder tensors in `fn`.
+    _ = scan(fn, init, xs)
```
[the remark this answers is not in the document]
Comment (tengyifei) [reply test/scan/test_scan.py:638]: Added a test.

File: torch_xla/core/xla_builder.py [hunk 7]
```diff
@@ -40,6 +40,21 @@ class Type:
...
 
+_PT_XLA_TYPE_MAP = {
+    torch.float32: Type.F32,
+    torch.float64: Type.F64,
+    torch.bfloat16: Type.BF16,
+    torch.float16: Type.F16,
+    torch.uint8: Type.U8,
+    torch.int8: Type.S8,
+    torch.int16: Type.S16,
+    torch.int32: Type.S32,
```
[the remark this answers is not in the document]
Comment (tengyifei) [reply torch_xla/core/xla_builder.py:51]: Added mapping for unsigned.

Good catch. I think the existing map has a bug. I'll file one: https://github.com/pytorch/xla/issues/8798

**Review by rpsilva-aws — APPROVED** [review]

File: torch_xla/experimental/scan.py [hunk 8]
```diff
@@ -558,8 +558,8 @@ def fn(carry, x):
...
-    return torch.empty(
-        v.size(), dtype=v.dtype).to(device).requires_grad_(v.requires_grad)
+    t = xb.create_placeholder_tensor(v.shape, v.dtype)
+    return t.requires_grad_(v.requires_grad)
```
Comment (zpcore) [inline torch_xla/experimental/scan.py:562]: I saw sharding issues when try scan recently. Not sure if we need any device info for placeholder tensor.

Comment (rpsilva-aws) [reply torch_xla/experimental/scan.py:562]: Could you create an issue with more details / repro?
